python/ray/data/dataset.py

```diff
@@ -2525,29 +2532,50 @@ def repeat(self, times: Optional[int] = None) -> "DatasetPipeline[T]":
         """
         from ray.data.dataset_pipeline import DatasetPipeline
 
+        if self._plan._is_read_stage():
```

Comment:
This part is "rewriting" the LazyBlockList into a concrete BlockList of read metadata, which is then executed by a OneToOne op. This allows fusion of read stages in pipelines with subsequent transformations.